### ISSUE TITLE
E-Book support

### ISSUE THREAD

**Issue by gh0sti** [rasmuslos/ShelfPlayer#35]
![image](https://github.com/rasmuslos/ShelfPlayer/assets/4202657/2ff406fe-331b-45c4-8383-a3c90a0a6d23)

Was looking at my comic/ebook library and noticed that it has play clock time and probably needs to be logically looked at how to present and manage ebooks and comics. Are you looking to do something like komga to stream pages to the client?

**Comment by gh0sti** [comment]
I would maybe make this a feature dev flag to disable for the time being if ebooks and comics aren't going to be supported for the time being.

**Comment by rasmuslos** [comment]
I would wait until ebooks / comics are fully supported in ABS until I consider implementing them. But I am not sure if they will ever be supported, because I would consider them out of scope for an app primarily aimed at playing audio

**Comment by gh0sti** [comment]
> I would wait until ebooks / comics are fully supported in ABS until I consider implementing them. But I am not sure if they will ever be supported, because I would consider them out of scope for an app primarily aimed at playing audio

They actually are supported on ABS

"Basic ebook support and ereader
Epub, pdf, cbr, cbz
Send ebook to device (i.e. Kindle)"

https://github.com/advplyr/audiobookshelf?tab=readme-ov-file#about

**Comment by rasmuslos** [comment]
Well as far as I can tell there is an API endpoint to get the pub / pdf / ... file, but not much more. This is very likely to change and become a more fully fledged solution.

I will reopen this to have something to go back to, but not until the ebook support is more than just basic

**Comment by lkjfdsaofmc** [comment]
If possible when EBook support is looked at again, it would be nice to have a feature to simply download the ebook file (such as epub) directly in the app in a way that allows you to then share it out the same as if you access ABS via a web browser on your phone.  This feature allows you to share it to something such as a kindle e-reader which would be a more convenient method for loading my ebooks to an e-reader than through my computer or phones web browser.

**Comment by rasmuslos** [comment]
> If possible when EBook support is looked at again, it would be nice to have a feature to simply download the ebook file (such as epub) directly in the app in a way that allows you to then share it out the same as if you access ABS via a web browser on your phone. This feature allows you to share it to something such as a kindle e-reader which would be a more convenient method for loading my ebooks to an e-reader than through my computer or phones web browser.

This would actually be a good (temporary) solution. I have looked at some other apps like `Smart Comic Reader` and looked for some libraries to handle the actual comic rendering and it seems like it would have to reinvent the wheel for this to work natively in ShelfPlayer. But I am pretty sure there is no way to sync progress back to ShelfPlayer...

**Comment by gh0sti** [comment]
> > If possible when EBook support is looked at again, it would be nice to have a feature to simply download the ebook file (such as epub) directly in the app in a way that allows you to then share it out the same as if you access ABS via a web browser on your phone. This feature allows you to share it to something such as a kindle e-reader which would be a more convenient method for loading my ebooks to an e-reader than through my computer or phones web browser.
> 
> This would actually be a good (temporary) solution. I have looked at some other apps like `Smart Comic Reader` and looked for some libraries to handle the actual comic rendering and it seems like it would have to reinvent the wheel for this to work natively in ShelfPlayer. But I am pretty sure there is no way to sync progress back to ShelfPlayer...

Looking at the current commits in ABS they added comic book support.

**Comment by rasmuslos** [comment]
Also putting this here for future reference: https://developer.apple.com/documentation/uikit/uidocumentinteractioncontroller

**Comment by lkjfdsaofmc** [comment]
An e-book can exist in ABS without an audiobook, but ABS doesn't read much of the metadata from the epub files beyond title.  My work-around thus far has been to create a 1 second blank audio file to hold the metadata for the e-book and keep that with the epub so ABS thinks there's an audiobook, but if you used online matching/metadata sources (which I have turned off completely) it would probably be fine having epub's alone.

**Comment by rasmuslos** [comment]
To give you some context why this takes so long: to use the document interaction controller and present a menu you have to specify the bounds and a view, but both are UIKit components. The app is written in SwiftUI, so I have to create an `UIViewRepresentable`, but I don't really know how this works and how to get the bounds.

**Comment by danielraffel** [comment]
It looks like epub support is enabled in the official client apps. Are there plans to add support soon?

https://github.com/advplyr/audiobookshelf-app

For now, I've disabled indexing non-audio files since they don't render nicely in this client (I get a play icon which shouldn't appear.) Thanks for making a nice, clean app!

**Comment by rasmuslos** [comment]
For now ebooks will be filtered, so they don't show up in your library. I check the `numAudioFiles` property for this, not ideal, but it should work until I find a better way to do this. The problem is that audiobooks and ebooks are virtually the same thing in ABS, but in ShelfPlayer they will most likely be their own thing.

**Comment by rasmuslos** [comment]
[This API](https://developer.apple.com/documentation/swiftui/view/fileexporter(ispresented:document:contenttype:defaultfilename:oncompletion:)-32vwk) should allow ShelfPlayer to export documents
